syntax = "proto3";

package election;

service Election {
    rpc ServeClient(ServeClientArgs) returns (ServeClientReply) {}
    rpc RequestVote(RequestVoteRequest) returns (RequestVoteResponse) {}
    rpc AppendEntries(RequestAppendEntries) returns (ResponseAppendEntries) {}
}

message ServeClientArgs {
    string Request = 1;
}

message ServeClientReply {
    string Data = 1;
    string LeaderID = 2;
    bool Success = 3;
}

message RequestVoteRequest {
    int32 term = 1;
    int32 candidate_id = 2;
    int32 lastLogIndex = 3;
    int32 lastLogTerm = 4;
}

message RequestVoteResponse {
    int32 term = 1;
    bool vote_granted = 2;
    int32 previousleaderlease = 4;
}

message RequestAppendEntries {
    int32 term = 1;
    int32 leaderId = 2;
    int32 prevLogIndex = 3;
    int32 prevLogTerm = 4;
    int32 leaderCommit = 5;
    repeated string entries = 6;
    int32 leaseintervalduration = 7;
}

message ResponseAppendEntries {
    int32 term = 1;
    int32 ack = 2;
    bool success = 3;
}
